src/compiler/checker.ts

```diff
@@ -3527,15 +3527,20 @@ namespace ts {
         }
 
         function getTypeNamesForErrorDisplay(left: Type, right: Type): [string, string] {
-            let leftStr = typeToString(left);
-            let rightStr = typeToString(right);
+            let leftStr = hasNoContextSymbolvalueDeclaration(left) ? typeToString(left, left.symbol.valueDeclaration) : typeToString(left);
+            let rightStr = hasNoContextSymbolvalueDeclaration(right) ? typeToString(right, right.symbol.valueDeclaration) : typeToString(right);
             if (leftStr === rightStr) {
                 leftStr = typeToString(left, /*enclosingDeclaration*/ undefined, TypeFormatFlags.UseFullyQualifiedType);
                 rightStr = typeToString(right, /*enclosingDeclaration*/ undefined, TypeFormatFlags.UseFullyQualifiedType);
             }
             return [leftStr, rightStr];
         }
 
+        function hasNoContextSymbolvalueDeclaration(type: Type): boolean {

```

Comment:
This only operates on `type.symbol`, so can just take the `symbol` directly. Suggested name `symbolValueDeclarationIsContextSensitive` and invert the return value. Implementation then is just
```
   return symbol && symbol.valueDeclaration && isExpression(symbol.valueDeclaration) && isContextSensitive(symbol.valueDeclaration);
```

Reply:
Thank you for the review. Since the behavior changes when the return value is inverted, I changed your suggestion a bit.